File: travel_expense_backend/src/modules/audit/entities/audit-log.entity.ts
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, ManyToOne, JoinColumn } from 'typeorm';
import { User } from '../../users/entities/user.entity';

@Entity('audit_logs')
export class AuditLog {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ nullable: true })
  userId: string;

  @ManyToOne(() => User, { nullable: true, onDelete: 'SET NULL' })
  @JoinColumn({ name: 'userId' })
  user: User;

  @Column()
  action: string; // E.g., 'SSO_LOGIN', 'TRAVEL_SUBMIT', 'APPROVAL_APPROVED', 'CLAIM_REIMBURSED'

  @Column()
  entityName: string; // E.g., 'TravelRequest', 'ExpenseClaim', etc.

  @Column({ nullable: true })
  entityId: string;

  @Column({ type: 'json', nullable: true })
  metadata: Record<string, any>;

  @CreateDateColumn()
  timestamp: Date;
}

File: travel_expense_app/lib/features/audit/domain/entities/audit_log_entity.dart
import 'package:equatable/equatable.dart';

class AuditLogEntity extends Equatable {
  final String id;
  final String action;
  final String entityName;
  final String entityId;
  final String? userId;
  final Map<String, dynamic>? metadata;
  final DateTime timestamp;

  const AuditLogEntity({
    required this.id,
    required this.action,
    required this.entityName,
    required this.entityId,
    this.userId,
    this.metadata,
    required this.timestamp,
  });

  @override
  List<Object?> get props => [id, action, entityName, entityId, userId, metadata, timestamp];
}
